contracts/libraries/OracleLibrary.sol

```diff
@@ -76,4 +76,30 @@ library OracleLibrary {
 
         return uint32(block.timestamp) - observationTimestamp;
     }
+
+    /// @notice Given a pool, it returns the tick value as of the start of the current block
+    /// @param pool Address of Uniswap V3 pool
+    /// @return The tick that the pool was in at the start of the current block
+    function getBlockStartingTick(address pool) internal view returns (int24) {
+        (, int24 tick, uint16 observationIndex, uint16 observationCardinality, , , ) = IUniswapV3Pool(pool).slot0();
+
+        // 2 observations are needed to reliably calculate the block starting tick
+        require(observationCardinality > 1, 'NEO');
+
+        // If the latest observation occurred in the past, then no tick-changing trades have happened in this block
+        // therefore the tick in `slot0` is the same as at the beginning of the current block.
+        // We don't need to check if this observation is initialized - it is guaranteed to be.
+        (uint32 observationTimestamp, int56 tickCumulative, , ) = IUniswapV3Pool(pool).observations(observationIndex);
+        if (observationTimestamp != uint32(block.timestamp)) {
+            return tick;
+        }
+
+        uint256 prevIndex = (observationIndex + observationCardinality - 1) % observationCardinality;
```

Comment:
ooh bit of a scary one, i think we need to upcast observationIndex before performing any operations, then downcast again after the mod?

Reply:
yes great catch, tested this in remix to double check